XMonad/Layout/PerScreen.hs

```diff
@@ -1,75 +1,119 @@
-{-# LANGUAGE FlexibleInstances, MultiParamTypeClasses #-}
+{-# LANGUAGE FlexibleInstances, MultiParamTypeClasses, LambdaCase #-}
 
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  XMonad.Layout.PerScreen
--- Description :  Configure layouts based on the width of your screen.
--- Copyright   :  (c) Edward Z. Yang
+-- Description :  Configure layouts based on the screen rectangle.
+-- Copyright   :  (c) Brandon S. Allbery KF8NH
 -- License     :  BSD-style (see LICENSE)
 --
--- Maintainer  :  <ezyang@cs.stanford.edu>
+-- Maintainer  :  <allbery.b@gmail.com>
 -- Stability   :  unstable
 -- Portability :  unportable
 --
--- Configure layouts based on the width of your screen; use your
--- favorite multi-column layout for wide screens and a full-screen
--- layout for small ones.
+-- Configure layouts based on the screen rectangle passed to the layout.
+-- This gives you true per-screen functionality.
+--
+-- The old PerScreen is now X.L.ByWidth. We re-export it deprecated for
+-- backward compatibility.
 -----------------------------------------------------------------------------
 
 module XMonad.Layout.PerScreen
     ( -- * Usage
       -- $usage
-      PerScreen,
+      OnScreen,
+      onScreen,
+      onScreens,
+      -- * Deprecated
+      -- $deprecated
+      BW.PerScreen,
       ifWider
     ) where
 
 import XMonad
 import qualified XMonad.StackSet as W
 
-import XMonad.Prelude (fromMaybe)
+import XMonad.Prelude (fromMaybe, fi)
+
+import qualified XMonad.Layout.ByWidth as BW
+
+import Data.List (find)
 
 -- $usage
 -- You can use this module by importing it into your ~\/.xmonad\/xmonad.hs file:
 --
--- > import XMonad.Layout.PerScreen
+-- > import XMonad.Layout.OnScreen
 --
 -- and modifying your layoutHook as follows (for example):
 --
--- > layoutHook = ifWider 1280 (Tall 1 (3/100) (1/2) ||| Full) Full
+-- > layoutHook = onScreen 1 (Tall 1 (3/100) (1/2) ||| Full) Full
 --
 -- Replace any of the layouts with any arbitrarily complicated layout.
--- ifWider can also be used inside other layout combinators.
+-- 'onScreen' can also be used inside other layout combinators, although the
+-- result may be confusing.
+
+-- | Specify a layout to run on a given screen.
+onScreen :: (LayoutClass l1 a, LayoutClass l2 a)
+         => ScreenId -> l1 a -> l2 a -> OnScreen l1 l2 a
+onScreen s = onScreens [s]
+
+-- | Specify a layout to run on a list of screens.
+--   Note that this works by 'ScreenId'. It has a 'Num' instance, so literal
+--   screen numbers will work as expected, but if you use a binding you need
+--   to use the 'S' constructor.
+onScreens :: (LayoutClass l1 a, LayoutClass l2 a)
+          => [ScreenId] -> l1 a -> l2 a -> OnScreen l1 l2 a
+onScreens ss l1 l2 = OnScreen ss l1 l2 False -- @@@ is this right?
```

Comment:
> @@@ is this right?

Are you asking about the flag? I think so since as soon as runLayout is executed the flag will be set correctly

Reply:
That was what I figured but I can't recall if there are special cases where the default might be used inappropriately. But even if there were it would be on an inactive screen by definition, which is why I used `False`.

Reply:
I don't think there are cases where it might be used inappropriately. **But** we're also not using the flag enough. We need to `Hide` whenever the flag flips, and dispatch `ReleaseResources` to both layouts.

(Might be better to finish #582 and get #75 fixed for free, like I mentioned in another comment.)

Reply:
If we receive `Hide`, do I need to relay it to both layouts or just the active one?

Reply:
This should be done, although I cheated a bit on `ReleaseResources`: should I check for a new constructor being produced by the forwarded message?

Also, if we receive `Hide`, should it be relayed to both sublayouts or only the active one?

Reply:
I now have both fully implemented `ReleaseResources` and a direct implementation of `Hide` which sends to both sublayouts because it's a PITA to send only to the active one. Also I'm relying in a few places on `Hide` being idempotent, but we may have other issues if it's not anyway.